File: GH.Core/Services/DataService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using GH.Core.Models;
using MongoDB.Driver;
using GH.Core.Extensions;
using MongoDB.Bson;

namespace GH.Core.Services
{
    public class DataService : IDataService
    {
        private IMongoCollection<Country> _countryCollection;
        private IMongoCollection<City> _cityCollection;
        private IMongoCollection<Region> _regionCollection;

        public DataService()
        {
            var mongoDb = MongoContext.Db;
            _countryCollection = mongoDb.Countries;
            _cityCollection = mongoDb.Cities;
            _regionCollection = mongoDb.Regions;
        }

        public List<Country> GetAllCountry()
        {
            return _countryCollection.Find(new BsonDocument()).ToList();
        }

        public List<City> GetCityByCountry(string countryCode)
        {
            return _cityCollection.Find(t => t.CountryCode == countryCode).ToList();
        }

        public List<City> GetCityByRegion(string regionCode)
        {
            return _cityCollection.Find(t => t.RegionCode == regionCode).ToList();
        }

        public List<Region> GetRegionByCountry(string countryCode)
        {
            return _regionCollection.Find(t => t.Code == countryCode).ToList();
        }
    }
}

File: GH.Web/Areas/User/Scripts/Services/DataService.js
var _dataApp = getApp('DataModule', [], false);

_dataApp.factory('DataService', ['$http', '$q', function ($http, $q) {

    var getAllCountry = function () {
        var deferer = $q.defer();

        $http.get('/Api/Data/Country', {})
            .success(function(res) {
                deferer.resolve(res.Countries);
            }).error(function(errors, status) {
                __promiseHandler.Error(errors, status, deferer);
            });

        return deferer.promise;
    }

    var getCityByCountry = function (countryCode) {
        var deferer = $q.defer();

        $http.get('/Api/Data/Country/City', { params: { countryCode: countryCode, hideAjaxLoader: true } })
            .success(function(res) {
                deferer.resolve(res.Cities);
            }).error(function(errors, status) {
                __promiseHandler.Error(errors, status, deferer);
            });

        return deferer.promise;
    }

    var getCityByRegion = function (regionCode) {
        var deferer = $q.defer();

        $http.get('/Api/Data/Region/City', { params: { regionCode: regionCode, hideAjaxLoader: true } })
            .success(function(res) {
                deferer.resolve(res.Cities);
            }).error(function(errors, status) {
                __promiseHandler.Error(errors, status, deferer);
            });

        return deferer.promise;
    }

    var getRegionByCountry = function (countryCode) {
        var deferer = $q.defer();

        $http.get('/Api/Data/Country/Region', { params: { countryCode: countryCode, hideAjaxLoader: true } })
            .success(function(res) {
                deferer.resolve(res.Regions);
            }).error(function(errors, status) {
                __promiseHandler.Error(errors, status, deferer);
            });

        return deferer.promise;
    }

    return {
        getAllCountry: getAllCountry,
        getCityByCountry: getCityByCountry,
        getCityByRegion: getCityByRegion,
        getRegionByCountry: getRegionByCountry,
    }
}])